File: app/src/main/java/com/google/android/leanbacklauncher/notifications/NotificationCardView.java
package com.google.android.leanbacklauncher.notifications;

import android.animation.Animator;
import android.animation.AnimatorListenerAdapter;
import android.animation.ObjectAnimator;
import android.app.PendingIntent;
import android.content.Context;
import android.content.pm.PackageManager.NameNotFoundException;
import android.content.res.Resources;
import android.graphics.Bitmap;
import android.graphics.Bitmap.Config;
import android.graphics.Canvas;
import android.graphics.Rect;
import android.graphics.drawable.BitmapDrawable;
import android.graphics.drawable.ColorDrawable;
import android.graphics.drawable.Drawable;
import android.support.annotation.Keep;
import android.support.v17.leanback.widget.BaseCardView;
import android.support.v4.content.res.ResourcesCompat;
import android.util.AttributeSet;
import android.util.Log;
import android.view.View;
import android.view.ViewGroup.LayoutParams;
import android.widget.ImageView;
import android.widget.ImageView.ScaleType;
import android.widget.ProgressBar;
import android.widget.TextView;
import com.google.android.leanbacklauncher.DimmableItem;
import com.google.android.leanbacklauncher.R;
import com.google.android.leanbacklauncher.animation.ParticipatesInLaunchAnimation;
import com.google.android.leanbacklauncher.animation.ParticipatesInScrollAnimation;
import com.google.android.leanbacklauncher.animation.ViewDimmer;
import com.google.android.leanbacklauncher.animation.ViewDimmer.DimState;
import com.google.android.leanbacklauncher.animation.ViewFocusAnimator;
import com.google.android.leanbacklauncher.util.Util;
import com.google.android.tvrecommendations.TvRecommendation;

public class NotificationCardView extends BaseCardView implements DimmableItem, ParticipatesInLaunchAnimation, ParticipatesInScrollAnimation {
    private boolean mAnimationsEnabled;
    private boolean mAutoDismiss;
    private ImageView mBadgeImage;
    private float mBadgeImageAlpha;
    private ImageView mBadgeImageSelected;
    private PendingIntent mClickedIntent;
    private Rect mClipBounds;
    private int mColor;
    private TextView mContentView;
    protected ViewDimmer mDimmer;
    private final int mFocusAnimDuration;
    protected ViewFocusAnimator mFocusAnimator;
    private final int mImageHeight;
    private final int mImageMaxWidth;
    private final int mImageMinWidth;
    private ImageView mImageView;
    private View mInfoArea;
    private ColorDrawable mInfoBackground;
    private ObjectAnimator mMetaAnim;
    private int mMetaUnselectedHeight;
    private View mMetadataArea;
    private ProgressBar mProgBar;
    private String mRecGroup;
    private PrescaledLayout mSelectedMetadataContainer;
    private String mSignature;
    private TextView mSourceNameView;
    private TextView mTitleView;
    private String mWallpaperUri;

    public NotificationCardView(Context context) {
        this(context, null);
    }

    public NotificationCardView(Context context, AttributeSet attrs) {
        this(context, attrs, 0);
    }

    public NotificationCardView(Context context, AttributeSet attrs, int defStyle) {
        super(context, attrs, defStyle);
        this.mAnimationsEnabled = true;
        Resources res = getResources();
        this.mImageMinWidth = res.getDimensionPixelOffset(R.dimen.notif_card_img_min_width);
        this.mImageMaxWidth = res.getDimensionPixelOffset(R.dimen.notif_card_img_max_width);
        this.mImageHeight = res.getDimensionPixelOffset(R.dimen.notif_card_img_height);
        this.mFocusAnimDuration = res.getInteger(R.integer.notif_card_metadata_animation_duration);
        this.mMetaUnselectedHeight = res.getDimensionPixelOffset(R.dimen.notif_card_info_height);
        this.mBadgeImageAlpha = res.getFraction(R.fraction.badge_icon_alpha, 1, 1);
        this.mDimmer = new ViewDimmer(this);
        this.mFocusAnimator = createNewFocusAnimator();
    }

    protected void onFinishInflate() {
        super.onFinishInflate();
        this.mImageView = (ImageView) findViewById(R.id.art_work);
        this.mMetadataArea = findViewById(R.id.metadata);
        this.mSelectedMetadataContainer = (PrescaledLayout) findViewById(R.id.selected_metadata_container);
        this.mInfoArea = findViewById(R.id.info_field);
        this.mTitleView = (TextView) findViewById(R.id.title_text);
        this.mContentView = (TextView) findViewById(R.id.content_text);
        this.mSourceNameView = (TextView) findViewById(R.id.source_name);
        this.mBadgeImage = (ImageView) findViewById(R.id.badge);
        this.mBadgeImageSelected = (ImageView) findViewById(R.id.badge_selected);
        this.mProgBar = (ProgressBar) findViewById(R.id.progress_bar);
        Drawable cardBkg = getBackground();
        this.mColor = ResourcesCompat.getColor(getResources(), R.color.notif_background_color, null);
        this.mInfoBackground = new ColorDrawable(this.mColor);
        this.mInfoArea.setBackground(this.mInfoBackground);
        this.mDimmer.addDimTarget(this.mImageView);
        this.mDimmer.addDimTarget(this.mTitleView);
        this.mDimmer.addDimTarget(this.mContentView);
        this.mDimmer.addDimTarget(this.mSourceNameView);
        this.mDimmer.addDimTarget(cardBkg);
        this.mDimmer.addDesatDimTarget(this.mBadgeImage);
        this.mDimmer.addDesatDimTarget(this.mBadgeImageSelected);
        this.mDimmer.addDimTarget(this.mInfoBackground);
        this.mDimmer.addDimTarget(this.mProgBar.getProgressDrawable());
        this.mDimmer.setDimLevelImmediate();
        setClipToOutline(true);
    }

    public void setColor(int color) {
        this.mColor = color;
        if (this.mColor != 0) {
            this.mInfoBackground.setColor(this.mColor);
        } else {
            this.mInfoBackground.setColor(ResourcesCompat.getColor(getResources(), R.color.notif_background_color, null));
        }
    }

    public void setMainImage(Drawable image) {
        this.mImageView.setImageDrawable(image);
    }

    public void setTitleText(CharSequence text) {
        this.mTitleView.setText(text);
    }

    public void setContentText(CharSequence text) {
        this.mContentView.setText(text);
    }

    public void setSourceName(CharSequence text) {
        this.mSourceNameView.setText(text);
    }

    public void setBadgeImage(Drawable image) {
        if (image != null) {
            image = image.mutate();
            this.mBadgeImage.setImageDrawable(getDeselectedBadgeIcon(image));
            this.mBadgeImage.setVisibility(0);
            this.mBadgeImageSelected.setImageDrawable(image);
            this.mBadgeImageSelected.setVisibility(0);
            return;
        }
        this.mBadgeImageSelected.setVisibility(8);
        this.mBadgeImage.setVisibility(8);
    }

    private Drawable getDeselectedBadgeIcon(Drawable image) {
        int iconWidth = image.getIntrinsicWidth();
        int iconHeight = image.getIntrinsicHeight();
        image.setBounds(0, 0, iconWidth, iconHeight);
        image.setFilterBitmap(true);
        image.setAlpha((int) (255.0f * this.mBadgeImageAlpha));
        Bitmap bitmap = Bitmap.createBitmap(iconWidth, iconHeight, Config.ARGB_8888);
        image.draw(new Canvas(bitmap));
        return new BitmapDrawable(getResources(), bitmap);
    }

    protected ViewFocusAnimator createNewFocusAnimator() {
        return new ViewFocusAnimator(this);
    }

    public void setWallpaperUri(String uri) {
        if (Util.isContentUri(uri)) {
            this.mWallpaperUri = uri;
            return;
        }
        this.mWallpaperUri = null;
        if (uri != null) {
            Log.w("NotificationCardView", "Invalid Content URI provided for recommendation background: " + uri);
        }
    }

    public String getWallpaperUri() {
        return this.mWallpaperUri;
    }

    protected void setSignature(String signature) {
        this.mSignature = signature;
    }

    public String getSignature() {
        return this.mSignature;
    }

    public void setProgressShown(boolean shown) {
        if (this.mProgBar != null) {
            this.mProgBar.setVisibility(shown ? 0 : 4);
        }
    }

    private static void setViewWidth(View v, int width) {
        LayoutParams p = v.getLayoutParams();
        if (p.width != width) {
            p.width = width;
            v.setLayoutParams(p);
        }
    }

    private static void setViewHeight(View v, int height) {
        LayoutParams p = v.getLayoutParams();
        if (p.height != height) {
            p.height = height;
            v.setLayoutParams(p);
        }
    }

    public void setClickedIntent(PendingIntent intent) {
        this.mClickedIntent = intent;
    }

    public PendingIntent getClickedIntent() {
        return this.mClickedIntent;
    }

    public void setAutoDismiss(boolean auto) {
        this.mAutoDismiss = auto;
    }

    public void setRecommendationGroup(String tag) {
        this.mRecGroup = tag;
    }

    public String getRecommendationGroup() {
        return this.mRecGroup;
    }

    public int getLaunchAnimationColor() {
        return this.mColor;
    }

    public void setDimState(DimState dimState, boolean immediate) {
        this.mDimmer.setDimState(dimState, immediate);
    }

    public void setSelected(boolean selected) {
        if (selected != isSelected()) {
            super.setSelected(selected);
            setMetaDataExpanded(selected);
        }
    }

    private void setNotificationImage(TvRecommendation rec) {
        try {
            Resources res = getContext().getPackageManager().getResourcesForApplication(rec.getPackageName());
            Drawable image = null;
            int width = -1;
            int height = -1;
            if (rec.getContentImage() != null) {
                image = new BitmapDrawable(res, rec.getContentImage());
                width = image.getIntrinsicWidth();
                height = image.getIntrinsicHeight();
            }
            if (rec.getWidth() > 0) {
                width = rec.getWidth();
            }
            if (rec.getHeight() > 0) {
                height = rec.getHeight();
            }
            this.mImageView.setScaleType(ScaleType.CENTER_CROP);
            setDimensions(width, height);
            setMainImage(image);
            setBadgeImage(res.getDrawable(rec.getBadgeIcon(), null));
        } catch (NameNotFoundException e) {
        }
    }

    protected void setDimensions(int imgWidth, int imgHeight) {
        int cardWidth;
        if (imgWidth <= 0 || imgHeight <= 0) {
            cardWidth = this.mImageMinWidth;
        } else {
            cardWidth = (int) (((float) imgWidth) / (((float) imgHeight) / ((float) this.mImageHeight)));
            if (cardWidth > this.mImageMaxWidth) {
                cardWidth = this.mImageMaxWidth;
            }
            if (cardWidth < this.mImageMinWidth) {
                cardWidth = this.mImageMinWidth;
            }
        }
        setViewWidth(this.mImageView, cardWidth);
        setViewHeight(this.mImageView, this.mImageHeight);
        if (((float) cardWidth) / ((float) this.mImageHeight) > 1.5f) {
            this.mTitleView.setMaxLines(1);
            this.mContentView.setMaxLines(1);
            return;
        }
        this.mTitleView.setMaxLines(2);
        this.mContentView.setMaxLines(2);
    }

    protected void onAttachedToWindow() {
        super.onAttachedToWindow();
        resetCardState();
    }

    protected void onDetachedFromWindow() {
        super.onDetachedFromWindow();
        resetCardState();
    }

    public void resetCardState() {
        boolean focus = hasFocus();
        super.setSelected(focus);
        if (this.mMetaAnim != null) {
            this.mMetaAnim.cancel();
            this.mMetaAnim = null;
        }
        clearAnimation();
        this.mDimmer.setDimLevelImmediate();
        this.mFocusAnimator.setFocusImmediate(focus);
        setMetaDataExpandedImmediate(focus);
        setAnimationsEnabled(true);
    }

    private void setMetaDataExpanded(final boolean expanded) {
        float f = 1.0f;
        if (this.mMetaAnim != null) {
            this.mMetaAnim.cancel();
            this.mMetaAnim = null;
        }
        if (this.mAnimationsEnabled && getVisibility() == 0 && hasWindowFocus() && isAttachedToWindow()) {
            float f2;
            this.mSelectedMetadataContainer.setVisibility(0);
            setMetadataOpenFraction(expanded ? 0.0f : 1.0f);
            String str = "metadataOpenFraction";
            float[] fArr = new float[2];
            if (expanded) {
                f2 = 0.0f;
            } else {
                f2 = 1.0f;
            }
            fArr[0] = f2;
            if (!expanded) {
                f = 0.0f;
            }
            fArr[1] = f;
            this.mMetaAnim = ObjectAnimator.ofFloat(this, str, fArr);
            this.mMetaAnim.setDuration((long) this.mFocusAnimDuration);
            this.mMetaAnim.addListener(new AnimatorListenerAdapter() {
                public void onAnimationStart(Animator animation) {
                    NotificationCardView.this.setHasTransientState(true);
                }

                public void onAnimationEnd(Animator animation) {
                    NotificationCardView.this.setHasTransientState(false);
                    NotificationCardView.this.setMetaDataExpandedImmediate(expanded);
                }
            });
            this.mMetaAnim.start();
            return;
        }
        setMetaDataExpandedImmediate(expanded);
    }

    private void setMetaDataExpandedImmediate(boolean expanded) {
        setMetadataAlphaAndPositionFraction(expanded ? 1.0f : 0.0f);
        setClipBounds(null);
        this.mSelectedMetadataContainer.setVisibility(expanded ? 0 : 8);
    }

    @Keep
    public void setMetadataOpenFraction(float fract) {
        setMetadataAlphaAndPositionFraction(fract);
        int height = this.mMetaUnselectedHeight + Math.round(((float) Math.max(0, this.mSelectedMetadataContainer.getContentHeight() - this.mMetaUnselectedHeight)) * fract);
        if (this.mClipBounds == null) {
            this.mClipBounds = new Rect();
        }
        this.mClipBounds.set(0, 0, getWidth(), (getHeight() - this.mMetadataArea.getHeight()) + height);
        setClipBounds(this.mClipBounds);
    }

    private void setMetadataAlphaAndPositionFraction(float fract) {
        int badgeTop = getRelativeTop(this.mBadgeImage, this.mMetadataArea);
        int badgeSelectedTop = getRelativeTop(this.mBadgeImageSelected, this.mMetadataArea);
        this.mSourceNameView.setAlpha(1.0f - fract);
        this.mBadgeImage.setAlpha(1.0f - fract);
        this.mBadgeImage.setTranslationY(((float) (badgeSelectedTop - badgeTop)) * fract);
        this.mInfoArea.getBackground().setAlpha((int) ((255.0f * fract) + 0.5f));
        this.mTitleView.setAlpha(fract);
        this.mContentView.setAlpha(fract);
        this.mBadgeImageSelected.setAlpha(fract);
        this.mBadgeImageSelected.setTranslationY(((float) ((-badgeSelectedTop) + badgeTop)) * (1.0f - fract));
    }

    private int getRelativeTop(View view, View ancestor) {
        if (view == ancestor) {
            return 0;
        }
        return getRelativeTop((View) view.getParent(), ancestor) + view.getTop();
    }

    protected void onFocusChanged(boolean gainFocus, int direction, Rect previouslyFocusedRect) {
        super.onFocusChanged(gainFocus, direction, previouslyFocusedRect);
        setSelected(gainFocus);
    }

    public void setAnimationsEnabled(boolean enabled) {
        this.mAnimationsEnabled = enabled;
        if (!(enabled || this.mMetaAnim == null || !this.mMetaAnim.isStarted())) {
            this.mMetaAnim.end();
        }
        this.mFocusAnimator.setEnabled(enabled);
        this.mDimmer.setAnimationEnabled(enabled);
    }
}

File: app/src/main/java/com/amazon/tv/leanbacklauncher/notifications/NotificationCardView.kt
package com.amazon.tv.leanbacklauncher.notifications

import android.animation.Animator
import android.animation.AnimatorListenerAdapter
import android.animation.ObjectAnimator
import android.app.PendingIntent
import android.content.Context
import android.content.pm.PackageManager
import android.graphics.Bitmap
import android.graphics.Canvas
import android.graphics.Rect
import android.graphics.drawable.BitmapDrawable
import android.graphics.drawable.ColorDrawable
import android.graphics.drawable.Drawable
import android.util.AttributeSet
import android.util.Log
import android.view.View
import android.widget.ImageView
import android.widget.ProgressBar
import android.widget.TextView
import androidx.annotation.Keep
import androidx.core.content.res.ResourcesCompat
import androidx.leanback.widget.BaseCardView
import com.amazon.tv.leanbacklauncher.DimmableItem
import com.amazon.tv.leanbacklauncher.R
import com.amazon.tv.leanbacklauncher.animation.ParticipatesInLaunchAnimation
import com.amazon.tv.leanbacklauncher.animation.ParticipatesInScrollAnimation
import com.amazon.tv.leanbacklauncher.animation.ViewDimmer
import com.amazon.tv.leanbacklauncher.animation.ViewDimmer.DimState
import com.amazon.tv.leanbacklauncher.animation.ViewFocusAnimator
import com.amazon.tv.leanbacklauncher.util.Util.isContentUri
import com.amazon.tv.tvrecommendations.TvRecommendation
import kotlin.math.max
import kotlin.math.roundToInt

class NotificationCardView @JvmOverloads constructor(
    context: Context?,
    attrs: AttributeSet? = null,
    defStyle: Int = 0
) : BaseCardView(context, attrs, defStyle), DimmableItem, ParticipatesInLaunchAnimation,
    ParticipatesInScrollAnimation {
    private var mAnimationsEnabled = true
    private var mAutoDismiss = false
    private var mBadgeImage: ImageView? = null
    private val mBadgeImageAlpha: Float
    private var mBadgeImageSelected: ImageView? = null
    var clickedIntent: PendingIntent? = null
    private var mClipBounds: Rect? = null
    private var launchAnimationColor = 0
    private var mContentView: TextView? = null
    private var mDimmer: ViewDimmer
    private val mFocusAnimDuration: Int
    private var mFocusAnimator: ViewFocusAnimator
    private val mImageHeight: Int
    private val mImageMaxWidth: Int
    private val mImageMinWidth: Int
    private var mImageView: ImageView? = null
    private var mInfoArea: View? = null
    private var mInfoBackground: ColorDrawable? = null
    private var mMetaAnim: ObjectAnimator? = null
    private val mMetaUnselectedHeight: Int
    private var mMetadataArea: View? = null
    private var mProgBar: ProgressBar? = null
    var recommendationGroup: String? = null
    private var mSelectedMetadataContainer: PrescaledLayout? = null
    var signature: String? = null
        private set
    private var mSourceNameView: TextView? = null
    private var mTitleView: TextView? = null
    private var mWallpaperUri: String? = null
    override fun onFinishInflate() {
        super.onFinishInflate()
        mImageView = findViewById(R.id.art_work)
        mMetadataArea = findViewById(R.id.metadata)
        mSelectedMetadataContainer = findViewById(R.id.selected_metadata_container)
        mInfoArea = findViewById(R.id.info_field)
        mTitleView = findViewById(R.id.title_text)
        mContentView = findViewById(R.id.content_text)
        mSourceNameView = findViewById(R.id.source_name)
        mBadgeImage = findViewById(R.id.badge)
        mBadgeImageSelected = findViewById(R.id.badge_selected)
        mProgBar = findViewById(R.id.progress_bar)
        val cardBkg = background
        launchAnimationColor =
            ResourcesCompat.getColor(resources, R.color.notif_background_color, null)
        mInfoBackground = ColorDrawable(launchAnimationColor)
        mInfoArea?.background = mInfoBackground
        mDimmer.addDimTarget(mImageView)
        mDimmer.addDimTarget(mTitleView)
        mDimmer.addDimTarget(mContentView)
        mDimmer.addDimTarget(mSourceNameView)
        mDimmer.addDimTarget(cardBkg)
        mDimmer.addDesatDimTarget(mBadgeImage)
        mDimmer.addDesatDimTarget(mBadgeImageSelected)
        mDimmer.addDimTarget(mInfoBackground)
        mDimmer.addDimTarget(mProgBar?.progressDrawable)
        mDimmer.setDimLevelImmediate()
        clipToOutline = true
    }

    fun setColor(color: Int) {
        launchAnimationColor = color
        if (launchAnimationColor != 0) {
            mInfoBackground?.color = launchAnimationColor
        } else {
            mInfoBackground?.color =
                ResourcesCompat.getColor(resources, R.color.notif_background_color, null)
        }
    }

    fun setMainImage(image: Drawable?) {
        mImageView?.setImageDrawable(image)
    }

    fun setTitleText(text: CharSequence?) {
        mTitleView?.text = text
    }

    fun setContentText(text: CharSequence?) {
        mContentView?.text = text
    }

    fun setSourceName(text: CharSequence?) {
        mSourceNameView?.text = text
    }

    fun setBadgeImage(image: Drawable?) {
        var img = image
        if (img != null) {
            img = img.mutate()
            mBadgeImage?.setImageDrawable(getDeselectedBadgeIcon(img))
            mBadgeImage?.visibility = View.VISIBLE
            mBadgeImageSelected?.setImageDrawable(img)
            mBadgeImageSelected?.visibility = View.VISIBLE
            return
        }
        mBadgeImageSelected?.visibility = View.GONE
        mBadgeImage?.visibility = View.GONE
    }

    private fun getDeselectedBadgeIcon(image: Drawable): Drawable {
        val iconWidth = image.intrinsicWidth
        val iconHeight = image.intrinsicHeight
        image.setBounds(0, 0, iconWidth, iconHeight)
        image.isFilterBitmap = true
        image.alpha = (255.0f * mBadgeImageAlpha).toInt()
        val bitmap = Bitmap.createBitmap(iconWidth, iconHeight, Bitmap.Config.ARGB_8888)
        image.draw(Canvas(bitmap))
        return BitmapDrawable(resources, bitmap)
    }

    private fun createNewFocusAnimator(): ViewFocusAnimator {
        return ViewFocusAnimator(this)
    }

    var wallpaperUri: String?
        get() = mWallpaperUri
        set(uri) {
            if (isContentUri(uri)) {
                mWallpaperUri = uri
                return
            }
            mWallpaperUri = null
            if (uri != null) {
                Log.w(
                    "NotificationCardView",
                    "Invalid Content URI provided for recommendation background: $uri"
                )
            }
        }

    fun setProgressShown(shown: Boolean) {
        if (mProgBar != null) {
            mProgBar?.visibility =
                if (shown) View.VISIBLE else View.INVISIBLE
        }
    }

    fun setAutoDismiss(auto: Boolean) {
        mAutoDismiss = auto
    }

    override fun setDimState(dimState: DimState, immediate: Boolean) {
        mDimmer.setDimState(dimState, immediate)
    }

    override fun setSelected(selected: Boolean) {
        if (selected != isSelected) {
            super.setSelected(selected)
            setMetaDataExpanded(selected)
        }
    }

    private fun setNotificationImage(rec: TvRecommendation) {
        try {
            val res = context.packageManager.getResourcesForApplication(rec.packageName)
            var image: Drawable? = null
            var width = -1
            var height = -1
            if (rec.contentImage != null) {
                image = BitmapDrawable(res, rec.contentImage)
                width = image.getIntrinsicWidth()
                height = image.getIntrinsicHeight()
            }
            if (rec.width > 0) {
                width = rec.width
            }
            if (rec.height > 0) {
                height = rec.height
            }
            mImageView?.scaleType = ImageView.ScaleType.CENTER_CROP
            setDimensions(width, height)
            setMainImage(image)
            setBadgeImage(ResourcesCompat.getDrawable(res, rec.badgeIcon, null))
        } catch (e: PackageManager.NameNotFoundException) {
            e.printStackTrace()
        }
    }

    private fun setDimensions(imgWidth: Int, imgHeight: Int) {
        var cardWidth: Int
        if (imgWidth <= 0 || imgHeight <= 0) {
            cardWidth = mImageMinWidth
        } else {
            cardWidth =
                (imgWidth.toFloat() / (imgHeight.toFloat() / mImageHeight.toFloat())).toInt()
            if (cardWidth > mImageMaxWidth) {
                cardWidth = mImageMaxWidth
            }
            if (cardWidth < mImageMinWidth) {
                cardWidth = mImageMinWidth
            }
        }
        setViewWidth(mImageView, cardWidth)
        setViewHeight(mImageView, mImageHeight)
        if (cardWidth.toFloat() / mImageHeight.toFloat() > 1.5f) {
            mTitleView?.maxLines = 1
            mContentView?.maxLines = 1
            return
        }
        mTitleView?.maxLines = 2
        mContentView?.maxLines = 2
    }

    override fun onAttachedToWindow() {
        super.onAttachedToWindow()
        resetCardState()
    }

    override fun onDetachedFromWindow() {
        super.onDetachedFromWindow()
        resetCardState()
    }

    private fun resetCardState() {
        val focus = hasFocus()
        super.setSelected(focus)
        if (mMetaAnim != null) {
            mMetaAnim?.cancel()
            mMetaAnim = null
        }
        clearAnimation()
        mDimmer.setDimLevelImmediate()
        mFocusAnimator.setFocusImmediate(focus)
        setMetaDataExpandedImmediate(focus)
        setAnimationsEnabled(true)
    }

    private fun setMetaDataExpanded(expanded: Boolean) {
        var f = 1.0f
        if (mMetaAnim != null) {
            mMetaAnim?.cancel()
            mMetaAnim = null
        }
        if (mAnimationsEnabled && visibility == View.VISIBLE && hasWindowFocus() && isAttachedToWindow) {
            mSelectedMetadataContainer?.visibility = View.VISIBLE
            setMetadataOpenFraction(if (expanded) 0.0f else 1.0f)
            val str = "metadataOpenFraction"
            val fArr = FloatArray(2)
            val f2: Float = if (expanded) {
                0.0f
            } else {
                1.0f
            }
            fArr[0] = f2
            if (!expanded) {
                f = 0.0f
            }
            fArr[1] = f
            mMetaAnim = ObjectAnimator.ofFloat(this, str, *fArr)
            mMetaAnim?.duration = mFocusAnimDuration.toLong()
            mMetaAnim?.addListener(object : AnimatorListenerAdapter() {
                override fun onAnimationStart(animation: Animator) {
                    setHasTransientState(true)
                }

                override fun onAnimationEnd(animation: Animator) {
                    setHasTransientState(false)
                    setMetaDataExpandedImmediate(expanded)
                }
            })
            mMetaAnim?.start()
            return
        }
        setMetaDataExpandedImmediate(expanded)
    }

    private fun setMetaDataExpandedImmediate(expanded: Boolean) {
        setMetadataAlphaAndPositionFraction(if (expanded) 1.0f else 0.0f)
        clipBounds = null
        mSelectedMetadataContainer?.visibility =
            if (expanded) View.VISIBLE else View.GONE
    }

    @Keep
    fun setMetadataOpenFraction(fract: Float) {
        setMetadataAlphaAndPositionFraction(fract)
        val height = mMetaUnselectedHeight + (max(
            0,
            mSelectedMetadataContainer!!.contentHeight - mMetaUnselectedHeight
        ).toFloat() * fract).roundToInt()
        if (mClipBounds == null) {
            mClipBounds = Rect()
        }
        mClipBounds!![0, 0, width] = getHeight() - mMetadataArea!!.height + height
        clipBounds = mClipBounds
    }

    private fun setMetadataAlphaAndPositionFraction(fract: Float) {
        val badgeTop = getRelativeTop(mBadgeImage, mMetadataArea)
        val badgeSelectedTop = getRelativeTop(mBadgeImageSelected, mMetadataArea)
        mSourceNameView!!.alpha = 1.0f - fract
        mBadgeImage!!.alpha = 1.0f - fract
        mBadgeImage!!.translationY = (badgeSelectedTop - badgeTop).toFloat() * fract
        mInfoArea!!.background.alpha = (255.0f * fract + 0.5f).toInt()
        mTitleView!!.alpha = fract
        mContentView!!.alpha = fract
        mBadgeImageSelected!!.alpha = fract
        mBadgeImageSelected!!.translationY =
            (-badgeSelectedTop + badgeTop).toFloat() * (1.0f - fract)
    }

    private fun getRelativeTop(view: View?, ancestor: View?): Int {
        return if (view === ancestor) {
            0
        } else getRelativeTop(
            view!!.parent as View,
            ancestor
        ) + view.top
    }

    override fun onFocusChanged(gainFocus: Boolean, direction: Int, previouslyFocusedRect: Rect?) {
        super.onFocusChanged(gainFocus, direction, previouslyFocusedRect)
        isSelected = gainFocus
    }

    override fun setAnimationsEnabled(enabled: Boolean) {
        mAnimationsEnabled = enabled
        if (!(enabled || mMetaAnim == null || !mMetaAnim!!.isStarted)) {
            mMetaAnim?.end()
        }
        mFocusAnimator.setEnabled(enabled)
        mDimmer.setAnimationEnabled(enabled)
    }

    companion object {
        private fun setViewWidth(v: View?, width: Int) {
            val p = v!!.layoutParams
            if (p.width != width) {
                p.width = width
                v.layoutParams = p
            }
        }

        private fun setViewHeight(v: View?, height: Int) {
            val p = v!!.layoutParams
            if (p.height != height) {
                p.height = height
                v.layoutParams = p
            }
        }
    }

    init {
        val res = resources
        mImageMinWidth = res.getDimensionPixelOffset(R.dimen.notif_card_img_min_width)
        mImageMaxWidth = res.getDimensionPixelOffset(R.dimen.notif_card_img_max_width)
        mImageHeight = res.getDimensionPixelOffset(R.dimen.notif_card_img_height)
        mFocusAnimDuration = res.getInteger(R.integer.notif_card_metadata_animation_duration)
        mMetaUnselectedHeight = res.getDimensionPixelOffset(R.dimen.notif_card_info_height)
        mBadgeImageAlpha = res.getFraction(R.fraction.badge_icon_alpha, 1, 1)
        mDimmer = ViewDimmer(this)
        mFocusAnimator = createNewFocusAnimator()
    }
}